File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
     vector<vector<int>> res;
     queue<TreeNode*> traversal;

     if (!root) return res;

     traversal.push(root);

     while (!traversal.empty()){
        vector<int> level;
        int n = traversal.size();

        for (int i = 0; i < n; i++){
            TreeNode* cur = traversal.front();
            traversal.pop();
            if (cur){
                level.push_back(cur->val);
                traversal.push(cur->left);
                traversal.push(cur->right);
            }   
        }
        if (level.empty()) break;
        res.push_back(level); 
     }   
     return res;
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-1.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n = len(matrix)
        m = len(matrix[0])
        res = []

        up = 0
        down = n - 1
        left = 0
        right = m - 1

        while left <= right and up <= down:
            for i in range(left, right+1):
                res.append(matrix[up][i])
            up += 1

            for i in range(up, down+1):
                res.append(matrix[i][right])
            right -= 1

            if up <= down:
                for i in range(right, left-1, -1):
                    res.append(matrix[down][i])
                down -= 1

            if left <= right:
                for i in range(down, up-1, -1):
                    res.append(matrix[i][left])
                left += 1

        return res

